src/wayland/xdg_foreign/mod.rs

```diff
@@ -0,0 +1,533 @@
+//! let (shell_state, _, _) = xdg_shell_init(
+//!     &mut display,
+//!     |event: XdgRequest, dispatch_data| { /* handle the shell requests here */ },
+//!     None
+//! );
+//!
+//! let (foreign_state, _, _) = xdg_foreign_init(&mut display, shell_state.clone(), None);
+//!
+//! // Good to go!
+//! ```
+
+use crate::wayland::compositor;
+use crate::wayland::shell::is_toplevel_equivalent;
+use crate::wayland::shell::legacy::WL_SHELL_SURFACE_ROLE;
+use crate::wayland::shell::xdg::ShellState;
+use rand::distributions::{Alphanumeric, DistString};
+use std::ops::Deref;
+use std::sync::{Arc, Mutex};
+use wayland_protocols::unstable::xdg_foreign::v2::server::{
+    zxdg_exported_v2, zxdg_exporter_v2, zxdg_imported_v2, zxdg_importer_v2,
+};
+use wayland_server::protocol::wl_surface::WlSurface;
+use wayland_server::{Display, Filter, Global, Main};
+
+use super::shell;
+
+/// An invalid surface error generated when an invalid surface is exported.
+#[derive(thiserror::Error, Debug)]
+pub enum InvalidSurface {
+    #[allow(missing_docs)]
+    #[error("The surface was dead")]
+    Dead,
+
+    #[allow(missing_docs)]
+    #[error("The surface is not a toplevel surface")]
+    NotToplevel,
+}
+
+/// Manages all exported and imported surfaces.
+#[derive(Debug)]
+pub struct XdgForeignState {
+    log: ::slog::Logger,
+    shell: Arc<Mutex<ShellState>>,
+    exports: Vec<Export>,
+}
+
+impl XdgForeignState {
+    /// Returns a reference to an export using the handle of the export.
+    pub fn is_export_alive(&self, handle: String) -> bool {
+        self.exports.iter().any(|export| export.handle == handle)
+    }
+
+    /// Exports a surface and returns a newly created handle to refer to the surface.
+    pub fn export_surface(&mut self, surface: &WlSurface) -> Result<String, InvalidSurface> {
+        if let Some(client) = surface.as_ref().client() {
+            if !shell::is_toplevel_equivalent(surface)
+                && compositor::get_role(surface) != Some(WL_SHELL_SURFACE_ROLE)
+            {
+                return Err(InvalidSurface::NotToplevel);
+            }
+
+            if let Some(exported) = client.create_resource::<zxdg_exported_v2::ZxdgExportedV2>(1) {
+                let handle = Alphanumeric.sample_string(&mut rand::thread_rng(), 32);
+
+                // FIXME: Compositor originating exports belong to the client the surface was
+                // created by, this however means no destructors are run on said object.
+                // I have no real way to hook up a destructor without either recursively storing the XdgForeignState inside itself or somehow wrapping the inner value of
+
+                self.exports.push(Export {
+                    surface: surface.clone(),
+                    handle: handle.clone(),
+                    inner: exported.deref().clone(),
+                    imports: vec![],
+                    compositor_created: true,
+                });
+
+                return Ok(handle);
+            }
```

Comment:
No, that cannot do. You cannot create an object for a client without sending it to it via a protocol event specifically defined to create it. That will just completely mess up the protocol state because the client does not know about this object, and quickly cause a protocol error. We cannot export a surface of a client without it knowing about it.

From our discussions on the chatroom I though you were talking about exporting a pseudo-surface completely controlled by the compositor, so that clients could set it as parent *as if* it was a surface of an other client.